src/lib/blkalloc/append_blk_allocator.cpp

```diff
@@ -0,0 +1,180 @@
+    HS_REL_ASSERT_EQ(m_sb->version, append_blkalloc_sb_version, "Invalid version of AppendBlkAllocator metablk");
+}
+
+// alloc a single block;
+BlkAllocStatus AppendBlkAllocator::alloc(BlkId& bid) {
+    if (available_blks() < 1) {
+        COUNTER_INCREMENT(m_metrics, num_alloc_failure, 1);
+        LOGERROR("No space left to serve request nblks: 1, available_blks: {}", available_blks());
+        return BlkAllocStatus::SPACE_FULL;
+    }
+
+    bid.set(m_last_append_offset, 1, m_chunk_id);
+    ++m_last_append_offset;
+    --m_freeable_nblks;
+
+    set_dirty_offset();
+
+    COUNTER_INCREMENT(m_metrics, num_alloc, 1);
+    return BlkAllocStatus::SUCCESS;
+}
+
+//
+// For append blk allocator, the assumption is only one writer will append data on one chunk.
+// If we want to change above design, we can open this api for vector allocation;
+//
+BlkAllocStatus AppendBlkAllocator::alloc(blk_count_t nblks, const blk_alloc_hints& hint,
+                                         std::vector< BlkId >& out_bids) {
+    if (available_blks() < nblks) {
+        COUNTER_INCREMENT(m_metrics, num_alloc_failure, 1);
+        LOGERROR("No space left to serve request nblks: {}, available_blks: {}", nblks, available_blks());
+        return BlkAllocStatus::SPACE_FULL;
+    } else if (nblks > BlkId::max_blks_in_op()) {
+        // consumer(vdev) already handles this case.
+        COUNTER_INCREMENT(m_metrics, num_alloc_failure, 1);
+        LOGERROR("Can't serve request nblks: {} larger than max_blks_in_op: {}", nblks, BlkId::max_blks_in_op());
+        return BlkAllocStatus::FAILED;
+    }
+
+    // Push 1 blk to the vector which has all the requested nblks;
+    out_bids.emplace_back(m_last_append_offset, nblks, m_chunk_id);
+
+    m_last_append_offset += nblks;
+    m_freeable_nblks -= nblks;
+
+    set_dirty_offset();
+
+    COUNTER_INCREMENT(m_metrics, num_alloc, 1);
+
+    return BlkAllocStatus::SUCCESS;
+}
+
+//
+// Alternative design option:
+// Option-1. Everytime offset is updated in a single append_blk_allocator instance, update the dirty buffer into
+// VDevCPContext, which contains a list of (ThreadVector) dirty buffers from all the append_blk_allocator instances
+// (protected by RCU);
+//
+// And cp_flush flush every dirty buffer in that list to disk;
+//
+// Option-2. vdev cp_flush will call each individual chunk's blkallocator and each append_blk_allocator instance flush
+// its dirty buffer on demond;
+//
+// We are using option-2.
+//
+
+//
+// cp_flush is happening in different thread that is running concurrently as AppendBlkAllocator::alloc;
+//
+// 1. if clear_dirty_offset happens before offset is updated by alloc, but before set_dirty_offset, we are flushing
+// offset twice, which is fine.
+// 2. If clear_dirty_offset happens after set_dirty_offset, we are flushing offset ahead of log
+// store, which is also fine during recovery, it is consistent after log store flush completes;
```

Comment:
it is fine for append offset to be ahead, in either case, crash recovery or next cp, we will achieve consistency between the appendblkallocator and logstore.

Reply:
next cp will achieve consistency between them. If cp didn't happen before we crash, log replay will alloc blocks and we don't open for ready until all log entries are replayed.

Reply:
not sure if I understand correctly , when someone call the alloc(),   we alloc a blk , update these two counters to (N+1, K -1) , and before we even returning the blk to caller,  the CP comes and flush the counter. 

Questsions:
1.  who generate the log and when the log was generated?  I was thinking the caller will log after it gets the blk and finish some business logic?
2.  If we crash at this point and during recovery,  we got updated counter (N+1, K-1) , even we have the log and replay , arent we make it (N+2, K-2)?

Reply:
We are going to use option-1 since now it is not just an dirty atomic counter, but a buffer (can't garunteen atomicity while racing with io thread). But if it were still an automic counter during recovery/reply, we will need to do something to make sure the reply achieves final consistency, e.g. compare the alloc requests with existing persisted last_append_offset and if head, no-op, etc. But this is based on we were on option-2. 

Yes, your are right Log was generated by HO and converted by ReplDeva as was indicated by RaftBasedDesign Doc.